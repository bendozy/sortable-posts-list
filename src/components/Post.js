import React from 'react';
import PropTypes from 'prop-types';
import { UP, DOWN } from '../constants';

// React Flip need the child component to not be stateless
/* eslint-disable react/prefer-stateless-function */
export default class Post extends React.Component {
  render() {
    const {
      post: { title },
      index,
      post,
      lastIndex,
      movePost: movePostAction,
    } = this.props;

    return (
      <div className="Post bg-white shadow rounded mb-6 flex justify-between">
        <div className="Post-title py-6 px-2 text-md">{title}</div>
        <div className="Post-buttons flex flex-col pr-2 justify-center">
          {index !== 0 && (
            <button
              type="button"
              className="Post-button moveUp"
              onClick={() => movePostAction({ direction: UP, post, index })}
            >
              <i className="text-2xl fa fa-angle-up" />
            </button>
          )}
          {index !== lastIndex && (
            <button
              type="button"
              className="Post-button moveDown"
              onClick={() => movePostAction({ direction: DOWN, post, index })}
            >
              <i className="text-2xl fa fa-angle-down" />
            </button>
          )}
        </div>
      </div>
    );
  }
}

Post.propTypes = {
  post: PropTypes.shape({
    title: PropTypes.string.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  lastIndex: PropTypes.number.isRequired,
  movePost: PropTypes.func.isRequired,
};
