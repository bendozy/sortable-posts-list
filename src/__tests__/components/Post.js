import React from 'react';
import toJson from 'enzyme-to-json';
import Post from '../../components/Post';

const movePost = jest.fn();

describe('<Post />', () => {
  it('should render a <Action /> component', () => {
    const post = {
      title: 'test',
    };
    const wrapper = shallow(
      <Post
        movePost={movePost}
        post={post}
        index={1}
        lastIndex={4}
      />,
    );

    wrapper.find('.Post-button').first().simulate('click');
    wrapper.find('.Post-button').at(1).simulate('click');

    expect(wrapper.find('.Post')).toHaveLength(1);
    expect(wrapper.find('.Post-button')).toHaveLength(2);
    expect(wrapper.find('.Post-title').text()).toBe(post.title);
    expect(movePost.mock.calls.length).toBe(2);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render a only DOWN button at first index', () => {
    const post = {
      title: 'test',
    };
    const wrapper = shallow(
      <Post
        movePost={movePost}
        post={post}
        index={0}
        lastIndex={4}
      />,
    );

    expect(wrapper.find('.moveDown')).toHaveLength(1);
    expect(wrapper.find('.moveUp')).toHaveLength(0);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render a only UP button at last index', () => {
    const post = {
      title: 'test',
    };
    const wrapper = shallow(
      <Post
        movePost={movePost}
        post={post}
        index={4}
        lastIndex={4}
      />,
    );

    expect(wrapper.find('.moveDown')).toHaveLength(0);
    expect(wrapper.find('.moveUp')).toHaveLength(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
