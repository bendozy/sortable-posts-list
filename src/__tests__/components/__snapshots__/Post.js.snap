// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post /> should render a <Action /> component 1`] = `
<div
  className="Post bg-white shadow rounded mb-6 flex justify-between"
>
  <div
    className="Post-title py-6 px-2 text-md"
  >
    test
  </div>
  <div
    className="Post-buttons flex flex-col pr-2 justify-center"
  >
    <button
      className="Post-button moveUp"
      onClick={[Function]}
      type="button"
    >
      <i
        className="text-2xl fa fa-angle-up"
      />
    </button>
    <button
      className="Post-button moveDown"
      onClick={[Function]}
      type="button"
    >
      <i
        className="text-2xl fa fa-angle-down"
      />
    </button>
  </div>
</div>
`;

exports[`<Post /> should render a only DOWN button at first index 1`] = `
<div
  className="Post bg-white shadow rounded mb-6 flex justify-between"
>
  <div
    className="Post-title py-6 px-2 text-md"
  >
    test
  </div>
  <div
    className="Post-buttons flex flex-col pr-2 justify-center"
  >
    <button
      className="Post-button moveDown"
      onClick={[Function]}
      type="button"
    >
      <i
        className="text-2xl fa fa-angle-down"
      />
    </button>
  </div>
</div>
`;

exports[`<Post /> should render a only UP button at last index 1`] = `
<div
  className="Post bg-white shadow rounded mb-6 flex justify-between"
>
  <div
    className="Post-title py-6 px-2 text-md"
  >
    test
  </div>
  <div
    className="Post-buttons flex flex-col pr-2 justify-center"
  >
    <button
      className="Post-button moveUp"
      onClick={[Function]}
      type="button"
    >
      <i
        className="text-2xl fa fa-angle-up"
      />
    </button>
  </div>
</div>
`;
